keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,350 @@
+
+
+In this table "policy-reader" might be something like the "calico-controller" or any other cni agent
+that reads the API Servers network policy portfolio and responds to it over time.
+
+|         input policy         | policy-reader (old)    | client (new)        |   |   |
+|------------------------------|------------------------|---------------------|---|---|
+| nsNameSelector + podSelector | ANY ns, pod restricted | ns + pod restricted |   |   |
+| nsNameSelector               | invalid (empty peer)   | ns restricted       |   |   |
+| podSelector                  | ANY namespace          | ANY namespace       |   |   |
+
+In an older client (one which predated the addition of a new, "nsNameSelector" field, we find
+that there is a scenario (`ANY ns, pod restricted`), wherein an old interpretation of the API
+assumes that the ONLY restricting field is a pod label.  Thus, although the user intended to
+restrict all traffic to a namespace/pod combination, the policy implementer does something **much**
+more promiscuous.
+
+This is all avoidable if we could assume, in all cases, that namespaces have a default label.
+
+There may also be many other usecases where the a security or permissions boundary might be conveniently defined in terms
+of a namepsace name, and that such a definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this 
+can be used as selector by arbitrary components.
+
+### Non-Goals
+
+* Publishing name labels on all resources
+* Exposing arbitrary fields as selectors
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
+
+If the label is incorrect or changed, the apiserver fails to validate the object
+
+
+### User Stories (Optional)
+
+See https://github.com/kubernetes/enhancements/pull/2113/files for relevant user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is reserved. However, since there
+is no validation of reserved label prefixes, such namespaces will be invalidated
+post-upgrade, i.e. no further writes allowed. For this reason, we can do one alpha
+release where we log/event if we observe usage of this key that is not compatible.
+
+## Design Details
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+  if obj.Labels == nil {
+  	obj.Labels = map[string]string{}
+  }
+  
+  if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
```

Comment:
seems like you would want this unconditional.  if this isn't forced, then your selection isn't guaranteed.

Reply:
I made a similar comment about forcing the invariant in https://github.com/kubernetes/enhancements/pull/2162/files#r530463637

One benefit to overriding in strategy is that a warning can be added indicating the label on the incoming request was overridden.

Reply:
@liggitt  Assuming, qouting you from above :), that since the invariant can be forced on read anyways, just that we lose warnings, this is resolved ... 
```
The only thing stomping in defaults would not let us do would be to detect/warn if a user tried to set the label to some other value (since by the time we got to the strategy the conflicting user-specified value would be gone). That seems unlikely to be done accidentally, ....  not sure how much I care about that.
```

@deads2k if you feel very strongly that we **must** warn users about this, then LMK and we can revisit using strategy.go I think.  

Thanks!

Reply:
I think we can do:

defaults.go:
```
if gate-enabled {
    if obj.Labels == nil {
￼  	    obj.Labels = map[string]string{}
￼    }
    obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
}
```

Then, elsewhere (@liggitt - where can we send API warnings?):

```
if not gate-enabled {
    if obj,Labels != nil && obj.Labels[“kubernetes.io/metadata.name”] != obj.Name {
        API warning "the label 'kubernetes.io/metadata.name' is reserved and such uses will be over-written starting in v1.22"
    }
}
```

Is that fair?  Is it too generous?  It's unlikely to REALLY be an issue, but we just can't know...

@liggitt Remind me why it's better to just stomp the value rather than fail the API operation if a user tries to change it?

Reply:
> @liggitt - where can we send API warnings?

we can send them from strategy methods (PrepareForCreate / PrepareForUpdate / Validate / ValidateUpdate) and anything we plumb the request context to

> Is that fair? Is it too generous? It's unlikely to REALLY be an issue, but we just can't know...

I think that's too generous. Setting a reserved label with a pretty clear name (in my mind, at least) to a value that does not match the label does not seem like a reasonable thing to delay a release over.

> @liggitt Remind me why it's better to just stomp the value rather than fail the API operation if a user tries to change it?

Failing the request means:
1. the failure is not API-visible (since the write is disallowed)
2. we're depending on the rejected caller behaving well in response to the rejection (not hot-looping, reporting the failure in a discoverable way)
3. fixing the issue means fixing the caller, which could involve getting a new copy of some operator from the provider (the owner of the cluster where the call is rejected might not be the owner of the problematic writer).

In contrast, stomping the label has the same end-result of a controller that updates the `kubernetes.io/metadata.name` label asynchronously, which I think would be a fair thing to do (since we reserve ownership of kubernetes.io labels). There's still the hypothetical possibility of a component that was 1) setting this kubernetes.io label differently and 2) behaving poorly when the value later did not match it's set value, but given how specific the label name is, I do not expect that would happen accidentally.

Reply:
i think this is all set ?   we generally agree that we'll stomp.

Reply:
Re: precedent, this seems similar when we picked GA `kubernetes.io/os` and `kubernetes.io/arch` label keys, and started reconciling them for Nodes. We didn't sweep for user-set labels with different values and take multiple releases to phase in the new ones, allowing conflicting user-set values in the meantime. I think the `kubernetes.io` label prefix gives us reasonable leeway here.

> I do not feel like waiting one release is a particular hardship, given this has been unsolved for literally years.
> ...
> OK. I buy that argument.

I can't tell whether you still want to phase it in over multiple releases or not.

Reply:
i feel like we can decide to wait a release cycle, or not, irrespective of this specific PR, so ill mark this as resolved, and then follow on w / @thockin in the next sig-network meeting (tomorrow)

Reply:
I spoke to @liggitt offline and I am conceding.  I think we do not NEED to do a cycle of warnings.  We can just enable this.

Given the scope, we can probably just go straight to GA. ?